tests/test_workflow_dataset_status.py

```diff
@@ -0,0 +1,571 @@
+            n_session_participants - n_success_post_reorg
+        )
+        doughnut.loc[
+            doughnut[Manifest.col_session_id] == session_id, Doughnut.col_in_bids
+        ] = [True] * n_success_bids + [False] * (
+            n_session_participants - n_success_bids
+        )
+
+    session_participant_counts_df = pd.DataFrame(
+        data=zip(session_ids, participant_counts),
+        columns=["session_id", "participant_count"],
+    )
+
+    return Doughnut(doughnut), session_participant_counts_df
+
+
+def make_bagel(
+    n_participants=10,
+    session_ids=("BL", "M12"),
+    n_success_percent=50,
+    pipeline_configs=(
+        ("dcm2bids", "1.0.0", "prepare"),
+        ("dcm2bids", "1.0.0", "convert"),
+        ("fmriprep", "1.0.0", "default"),
+    ),
+    randomize_counts=False,
+) -> Bagel:
+
+    bagel, session_participant_counts_df = make_manifest(
+        n_participants, session_ids, randomize_counts
+    )
+
+    participant_ids = bagel["participant_id"].unique()
+    n_configs = len(pipeline_configs)
+    # add pipeline columns
+    _df_list = []
+    for config in pipeline_configs:
+        _df = bagel.copy()
+        _df[
+            [
+                Bagel.col_pipeline_name,
+                Bagel.col_pipeline_version,
+                Bagel.col_pipeline_step,
+            ]
+        ] = config
+        _df_list.append(_df)
+
+    bagel = Bagel(pd.concat(_df_list))
+
+    # add pipeline status columns
+    # always keep the first session as complete for all pipelines
+    bagel[[Bagel.col_status]] = "INCOMPLETE"
+    bagel.loc[bagel[Manifest.col_session_id] == session_ids[0], Bagel.col_status] = (
+        "SUCCESS"
+    )
+
+    # repeated participants for each pipeline config
+    participant_counts = [n_configs * len(participant_ids)]
+
+    # add the rest of the sessions
+    for session_id in session_ids[1:]:
+        n_session_participants = (
+            n_configs
+            * session_participant_counts_df[
+                session_participant_counts_df[Manifest.col_session_id] == session_id
+            ]["participant_count"].values[0]
+        )
+        n_success_pipeline = int(n_session_participants * n_success_percent / 100)
+        participant_counts.append(n_success_pipeline)
+
+        bagel.loc[bagel[Manifest.col_session_id] == session_id, Bagel.col_status] = [
+            "SUCCESS"
+        ] * n_success_pipeline + ["INCOMPLETE"] * (
+            n_session_participants - n_success_pipeline
+        )
+
+    session_participant_counts_df = pd.DataFrame(
+        data=zip(session_ids, participant_counts),
+        columns=["session_id", "participant_count"],
+    )
+
+    return bagel, session_participant_counts_df
+
+
+def make_status_df(
+    n_participants,
+    session_ids=("BL", "M12", "M24"),
+    pipeline_configs=(
+        ("dcm2bids", "1.0.0", "prepare"),
+        ("dcm2bids", "1.0.0", "convert"),
+        ("fmriprep", "1.0.0", "default"),
+    ),
```

Comment:
I think this function is implementing too much of the logic that is being used in the tested code. My preference would be to generate the `status_df` as directly as possible, e.g.

```python
def make_status_df(
    session_ids=("BL", "M12", "M24"),
    counts_by_col=(
        ("in_manifest", (20, 20, 19)),
        ("in_pre_reorg", (18, 20, 19)),
        ("in_post_reorg", (18, 18, 18)),
        ("in_bids", (18, 18, 18)),
        ("fmriprep\n1.0.0\ndefault", (0, 5, 5)),
        ("freesurfer\n1.0.0\ndefault", (18, 18, 18)),
    )
):
    df = pd.DataFrame(session_ids, columns=[Manifest.col_session_id])
    for col, counts in counts_by_col:
        df[col] = counts
    return df
```

Then for the test case, we could have parametrization like this:

```python
@pytest.mark.parametrize(
    ",session_ids,counts_by_col,emojis",
    [
        (
            ("BL", "M12"),
            (
                ("in_manifest", (20, 20)),
                ("in_pre_reorg", (18, 0)),
                ("in_post_reorg", (18, 0)),
                ("in_bids", (18, 0)),
                ("fmriprep\n1.0.0\ndefault", (20, 0)),
                ("freesurfer\n1.0.0\ndefault", (18, 0)),
            ),
            (
                {
                    "init\nrewards": [clipboard_emoji],
                    "curation\nrewards": [doughnut_emoji, broom_emoji],
                    "processing\nrewards": [bagel_emoji, party_popper_emoji],
                },
                {
                    "init\nrewards": [sad_cat_emoji],
                    "curation\nrewards": [],
                    "processing\nrewards": [],
                },
            ),
        ),
    ],
)

```

Reply:
I tried this at first, but it gets too annoying to manually adjust numbers in `status_df` when criteria for rewards is changed. I agree there is added complexity / logic in creating this test case, but I see the generative status_df and reductive reward operations as inverses of each other. To me this is different than duplicating the logic which would be problematic from testing perspective.

Reply:
I guess the main problem is that it is hard to be sure that the dataframe-generating helper function is itself correct. We don’t want to write tests for our tests…